### ISSUE TITLE
Internals::SvREFCNT inconsistent

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11741]
Migrated from [rt.perl.org#103222](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=103222) (status was 'resolved')

Searchable as RT103222$

**Comment by p5pRT** [comment]
### From @tonycoz

#### Created by @tonycoz



The Internals&#8203;::SvREFCNT\(\) function correctly adjusts the reference
count returned for the extra reference created when reading the
reference count\, but doesn't do so when setting the reference count&#8203;:

tony@&#8203;mars&#8203;:\.\.\./git/perl$ perl \-E '$x = 1; Internals&#8203;::SvREFCNT\($x\, 5\); say Internals&#8203;::SvREFCNT\($x\)'
4

&nbsp;    if \(items == 1\)
&nbsp;     XSRETURN\_IV\(SvREFCNT\(sv\) \- 1\); /\* Minus the ref created for us\. \*/
&nbsp;    else if \(items == 2\) \{
&nbsp;         /\* I hope you really know what you are doing\. \*/
&nbsp;     SvREFCNT\(sv\) = SvIV\(ST\(1\)\);
&nbsp;     XSRETURN\_IV\(SvREFCNT\(sv\)\);
&nbsp;    \}

This is easy enough to fix\, but is there CPAN or DarkPAN code that
relies upon the current behaviour?  \(grep\.cpan\.me was timing out when
I wrote this\.\)

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.15.4:

Configured by tony at Wed Nov  9 14:07:45 EST 2011.

Summary of my perl5 (revision 5 version 15 subversion 4) configuration:
  Commit id: 0765b2b82b120f4e57571368932aa4863aba62f6
  Platform:
    osname=linux, osvers=2.6.32-5-amd64, archname=x86_64-linux-thread-multi
    uname='linux mars 2.6.32-5-amd64 #1 smp mon oct 3 03:59:20 utc 2011 x86_64 gnulinux '
    config_args='-des -Dusedevel -Duseithreads -Dprefix=/home/tony/perl/blead -Uversiononly'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.4.5', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/../lib /usr/lib/../lib /lib /usr/lib /lib64 /usr/lib64
    libs=-lnsl -lgdbm -ldl -lm -lcrypt -lutil -lpthread -lc -lgdbm_compat
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=/lib/libc-2.11.2.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.11.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.15.4:
    /home/tony/perl/blead/lib/site_perl/5.15.4/x86_64-linux-thread-multi
    /home/tony/perl/blead/lib/site_perl/5.15.4
    /home/tony/perl/blead/lib/5.15.4/x86_64-linux-thread-multi
    /home/tony/perl/blead/lib/5.15.4
    .


Environment for perl 5.15.4:
    HOME=/home/tony
    LANG=en_AU.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/tony/perl/blead/bin:/home/tony/bin:/usr/local/bin:/usr/bin:/bin:/usr/games
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @Leont

On Wed\, Nov 9\, 2011 at 6&#8203;:40 AM\, Tony Cook \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> The Internals&#8203;::SvREFCNT\(\) function correctly adjusts the reference
> count returned for the extra reference created when reading the
> reference count\, but doesn't do so when setting the reference count&#8203;:
>
> tony@&#8203;mars&#8203;:\.\.\./git/perl$ perl \-E '$x = 1; Internals&#8203;::SvREFCNT\($x\, 5\); say Internals&#8203;::SvREFCNT\($x\)'
> 4
>
>    if \(items == 1\)
>         XSRETURN\_IV\(SvREFCNT\(sv\) \- 1\); /\* Minus the ref created for us\. \*/
>    else if \(items == 2\) \{
>         /\* I hope you really know what you are doing\. \*/
>         SvREFCNT\(sv\) = SvIV\(ST\(1\)\);
>         XSRETURN\_IV\(SvREFCNT\(sv\)\);
>    \}
>
> This is easy enough to fix\, but is there CPAN or DarkPAN code that
> relies upon the current behaviour?  \(grep\.cpan\.me was timing out when
> I wrote this\.\)

IMHO any code relying on setting the reference count is broken by
design\, and anything in Internals \(and I say that as someone who
depends on Internals\) may change on a whim\, so I'd say go ahead\.

Leon

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @tonycoz

On Tue\, Nov 08\, 2011 at 09&#8203;:40&#8203;:20PM \-0800\, Tony Cook wrote&#8203;:
> The Internals&#8203;::SvREFCNT\(\) function correctly adjusts the reference
> count returned for the extra reference created when reading the
> reference count\, but doesn't do so when setting the reference count&#8203;:
> 
> tony@&#8203;mars&#8203;:\.\.\./git/perl$ perl \-E '$x = 1; Internals&#8203;::SvREFCNT\($x\, 5\); say Internals&#8203;::SvREFCNT\($x\)'
> 4
> 
>     if \(items == 1\)
>      XSRETURN\_IV\(SvREFCNT\(sv\) \- 1\); /\* Minus the ref created for us\. \*/
>     else if \(items == 2\) \{
>          /\* I hope you really know what you are doing\. \*/
>      SvREFCNT\(sv\) = SvIV\(ST\(1\)\);
>      XSRETURN\_IV\(SvREFCNT\(sv\)\);
>     \}
> 
> This is easy enough to fix\, but is there CPAN or DarkPAN code that
> relies upon the current behaviour?  \(grep\.cpan\.me was timing out when
> I wrote this\.\)

The only match I can see on CPAN is Devel\-Valgrind\-Client\-0\.01\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

On Tue\, Nov 08\, 2011 at 09&#8203;:40&#8203;:20PM \-0800\, Tony Cook wrote&#8203;:
> The Internals&#8203;::SvREFCNT\(\) function correctly adjusts the reference
> count returned for the extra reference created when reading the
> reference count\, but doesn't do so when setting the reference count&#8203;:
> 
\.\.\.
> 
> This is easy enough to fix\, but is there CPAN or DarkPAN code that
> relies upon the current behaviour?  \(grep\.cpan\.me was timing out when
> I wrote this\.\)

Fixed in 28884311ae0933070ec776bd972da9d711a5184b\.

Tony

**Comment by p5pRT** [comment]
@tonycoz - Status changed from 'open' to 'resolved'
